x/accounts/extensions/feegrant/feegrant.go

```diff
@@ -0,0 +1,253 @@
+package feegrant
+
+import (
+	"context"
+	"time"
+
+	"cosmossdk.io/collections"
+	"cosmossdk.io/core/address"
+	"cosmossdk.io/core/appmodule"
+	corecontext "cosmossdk.io/core/context"
+	"cosmossdk.io/core/event"
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/x/accounts/accountstd"
+	"cosmossdk.io/x/accounts/extensions/feegrant/v1"
+	"cosmossdk.io/x/accounts/internal/implementation"
+	xfeegrant "cosmossdk.io/x/feegrant"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+type Dependencies = implementation.Dependencies
+
+var (
+	// todo: prefix conflicts when embedded in other account type. bump +100
+	PrefixGrants       = collections.NewPrefix(100)
+	PrefixGrantsExpiry = collections.NewPrefix(101)
+)
+
+type Feegrant struct {
+	env          appmodule.Environment
+	addressCodec address.Codec
+	feeAllowance collections.Map[sdk.AccAddress, v1.Grant]
+	// FeeAllowanceQueue key: expiration time+grantee | value: bool
+	feeAllowanceQueue collections.Map[collections.Pair[time.Time, sdk.AccAddress], bool]
+}
+
+func NewAccountExtension(d Dependencies, reg implementation.ProtoMsgHandlerRegistry) (*Feegrant, error) {
+	f := &Feegrant{
+		env:               d.Environment,
+		addressCodec:      d.AddressCodec,
+		feeAllowance:      collections.NewMap(d.SchemaBuilder, PrefixGrants, "fee_allowance", sdk.AccAddressKey, codec.CollValue[v1.Grant](d.LegacyStateCodec)),
+		feeAllowanceQueue: collections.NewMap(d.SchemaBuilder, PrefixGrantsExpiry, "fee_allowance_expiry", collections.PairKeyCodec(sdk.TimeKey, sdk.LengthPrefixedAddressKey(sdk.AccAddressKey)), collections.BoolValue),
+	}
+	implementation.RegisterExecuteHandler(reg, f.Init)
+	implementation.RegisterExecuteHandler(reg, f.GrantAllowance)
+	implementation.RegisterExecuteHandler(reg, f.UseGrantedFees)
```

Comment:
personal preference: I prefer all the wiring in the constructor. People would look here when doing copy paste for custom types. Methods can easily be missed when not required by an interface